chain/chain/src/trie_store_resharder.rs

```diff
@@ -0,0 +1,164 @@
+//! Logic for writing trie nodes with the appropriate shard prefix.
+//! This logic runs in parallel to chain processing.
+
+use crate::types::RuntimeAdapter;
+use near_chain_configs::{MutableConfigValue, ReshardingConfig};
+use near_primitives::{hash::CryptoHash, state};
+use near_store::{ShardUId, adapter::StoreAdapter, flat::ParentSplitParameters};
+use std::sync::Arc;
+
+pub struct TrieStoreResharder {
+    runtime: Arc<dyn RuntimeAdapter>,
+    /// Configuration for resharding.
+    resharding_config: MutableConfigValue<ReshardingConfig>,
+}
+
+impl TrieStoreResharder {
+    pub fn new(
+        runtime: Arc<dyn RuntimeAdapter>,
+        resharding_config: MutableConfigValue<ReshardingConfig>,
+    ) -> Self {
+        Self { runtime, resharding_config }
+    }
+
+    fn split_shard_task_impl(
+        &self,
+        shard_uid: ShardUId,
+        state_root: CryptoHash, // TODO: fix
+        parent_split_parameters: ParentSplitParameters,
+    ) -> Result<(), String> {
+        // Determines after how many bytes worth of key-values the process stops to commit changes
+        // and to check cancellation.
+        let batch_size = self.resharding_config.get().batch_size.as_u64() as usize;
+        // Delay between every batch.
+        let batch_delay = self.resharding_config.get().batch_delay.unsigned_abs();
+
+        // Iterate over the memtrie and write the new trie nodes to the new store.
+        let tries = self.runtime.get_tries();
+
+        let shard_uid = parent_split_parameters.left_child_shard; // TODO: fix
+        tries.get_trie_for_shard(shard_uid, state_root);
+        let store = self.runtime.store().trie_store();
+        let mut update = store.store_update();
+        // update.increment_refcount(shard_uid, hash, data);
+
+        todo!("unfinished");
+    }
+}
+
+/// All different states of task execution for [TrieStoreReshardingEventStatus].
+#[derive(Clone, Debug, Copy, Eq, PartialEq)]
+pub enum TaskExecutionStatus {
```

Comment:
What is your goal here? Do you mean to support node restarts? If not too wasteful can you first implement a simpler version where it is not supported and add it in a follow up PR? We can also discuss how to support restarts because it may be easier to just start from scratch than actually keeping track of the progress and resuming.

Reply:
Ah I think you're aiming to pause between batches -> flat storage resharder just sleeps there, can you do the same? Also the previous comment still applies -> can you first implement a simple version without batches and add batching later?